### PR TITLE
#1536 npm package generator with size

### LINKED ISSUE #1536: Implement support for generating test artifacts with a specified size for Npm
# Task Description

We need to be able to create test artifacts with a specified size.

# Classes of Interest

* `org.carlspring.strongbox.artifact.generator.NpmArtifactGenerator`
* `org.carlspring.strongbox.testing.artifact.NpmTestArtifact`

# Tasks

The following tasks will need to be carried out:
* [ ] Apply the necessary changes to the `org.carlspring.strongbox.artifact.generator.NpmArtifactGenerator`.
* [ ] Apply the necessary changes to the `org.carlspring.strongbox.testing.artifact.NpmTestArtifact`.
* [ ] Add tests for this.

# Task Relationships

This task:
* Is a sub-task of: #1534 
* Relates to: #1535 , #1537 , #1538 , #1540 

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @fuss86
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by wpater**
# Pull Request Description

Add support for generating NPM packeges with given size.
This pull request closes #1536 

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

**Comment by carlspring** [comment]
It would probably be best to apply you `TestFileUtils` change to the rest of the layout provider tests.

**Comment by wpater** [comment]
> It would probably be best to apply you `TestFileUtils` change to the rest of the layout provider tests.

I will do it in other tasks. In this one only for NPM and Maven, task for Maven is already closed.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java [hunk 1]
```diff
@@ -14,6 +15,7 @@
           EventsConfig.class,
           DataServiceConfig.class,
           CommonConfig.class,
+          CronTasksConfig.class,
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: Why is this necessary?

Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: Could not start without it. It's already added in other configurations, for example in `Maven2LayoutProviderTestConfig`. I don't have enough knowledge about whole project to answer with details. :(

Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: @fuss86 / @sbespalov : Is this required? Any advice?

Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: In `NpmRepositoryManagementStrategy`:
```
@Inject
private CronTaskDataService cronTaskDataService;
```
Without `CronTasksConfig` I am receiving `NoSuchBeanDefinitionException`.

Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: @ptirador @fuss86 @sbespalov Could you guys help with that?

File: strongbox-testing/strongbox-testing-core/src/main/java/org/carlspring/strongbox/util/TestFileUtils.java [hunk 2]
```diff
@@ -21,4 +30,50 @@ public static void deleteIfExists(File file)
...
+
+    public static void generateFile(ZipOutputStream zos,
+                                    long size,
+                                    String name)
+            throws IOException
+    {
+        ZipEntry ze = new ZipEntry(name);
+        zos.putNextEntry(ze);
+
+        RandomInputStream ris = new RandomInputStream(size);
```
Comment (fuss86) [inline strongbox-testing/strongbox-testing-core/src/main/java/org/carlspring/strongbox/util/TestFileUtils.java]: I know that you pasted in from `MavenArtifactGenerator `, but could you use here [The try-with-resources Statement](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) , please  ?

File: strongbox-testing/strongbox-testing-core/src/main/java/org/carlspring/strongbox/util/TestFileUtils.java [hunk 3]
```diff
@@ -21,4 +30,50 @@ public static void deleteIfExists(File file)
...
+    }
+
+    public static void generateFile(OutputStream bos,
+                                    long size)
+            throws IOException
+    {
+        bos.write("data = \"".getBytes(StandardCharsets.UTF_8));
+
+        OutputStream dataOut = Base64.getEncoder().wrap(bos);
+        RandomInputStream ris = new RandomInputStream(size);
```
Comment (fuss86) [inline strongbox-testing/strongbox-testing-core/src/main/java/org/carlspring/strongbox/util/TestFileUtils.java]: Same as above 
Use [The try-with-resources Statement](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) , please.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java (hunk 1 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: Yeah, it looks like `NpmLayoutProviderTestConfig` is not used nowhere else except `NpmLayoutProviderCronTasksTestConfig`. Because, as it looks like, we still need `CronTasksConfig` in tests from outside of the `cron tasks tests` I suggest to merge `@Import`s into `NpmLayoutProviderTestConfig` and use only `NpmLayoutProviderTestConfig` and drop `NpmLayoutProviderCronTasksTestConfig`

**Comment by fuss86** [comment]
Also, I'd suggest to use `bytesSize` instead of `size` around this whole PR. This way `dev` doesn't have to check what a `size` actually is.

What do think, @carlspring ?

**Comment by carlspring** [comment]
> Also, I'd suggest to use `bytesSize` instead of `size` around this whole PR. This way `dev` doesn't have to check what a `size` actually is.
> 
> What do think, @carlspring ?

I think you're right, @fuss86 .

@wpater : Would you mind making this change as well (and marking all the comments where you've already applied fixes as "resolved", if they have been resolved)? Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java (hunk 1 above)
Comment (wpater) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/test/java/org/carlspring/strongbox/config/NpmLayoutProviderTestConfig.java]: `NpmLayoutProviderCronTasksTestConfig` removed.

**Comment by wpater** [comment]
> > Also, I'd suggest to use `bytesSize` instead of `size` around this whole PR. This way `dev` doesn't have to check what a `size` actually is.
> > What do think, @carlspring ?
> 
> I think you're right, @fuss86 .
> 
> @wpater : Would you mind making this change as well (and marking all the comments where you've already applied fixes as "resolved", if they have been resolved)? Thanks! :)

Done.

**Comment by carlspring** [comment]
> > > Also, I'd suggest to use `bytesSize` instead of `size` around this whole PR. This way `dev` doesn't have to check what a `size` actually is.
> > > What do think, @carlspring ?
> > 
> > 
> > I think you're right, @fuss86 .
> > @wpater : Would you mind making this change as well (and marking all the comments where you've already applied fixes as "resolved", if they have been resolved)? Thanks! :)
> 
> Done.

Thank you, @wpater ! :)
